src/function/scalar/date/date_sub.cpp

```diff
@@ -0,0 +1,434 @@
+#include "duckdb/common/vector_operations/vector_operations.hpp"
+#include "duckdb/common/string_util.hpp"
+#include "duckdb/storage/statistics/numeric_statistics.hpp"
+
+namespace duckdb {
+
+struct DateSub {
+	struct MonthOperator {
+		template <class TA, class TB, class TR>
+		static inline TR Operation(TA start_ts, TB end_ts) {
+
+			if (start_ts > end_ts) {
+				return -MonthOperator::Operation<TA, TB, TR>(end_ts, start_ts);
+			}
+			// The number of complete months depends on whether end_ts is on the last day of the month.
+			date_t end_date;
+			dtime_t end_time;
+			Timestamp::Convert(end_ts, end_date, end_time);
+
+			int32_t yyyy, mm, dd;
+			Date::Convert(end_date, yyyy, mm, dd);
+			const auto end_days = Date::MonthDays(yyyy, mm);
+			if (end_days == dd) {
+				// Now check whether the start day is after the end day
+				date_t start_date;
+				dtime_t start_time;
+				Timestamp::Convert(start_ts, start_date, start_time);
+				Date::Convert(start_date, yyyy, mm, dd);
+				if (dd > end_days || (dd == end_days && start_time < end_time)) {
+					// Move back to the same time on the last day of the (shorter) end month
+					start_date = Date::FromDate(yyyy, mm, end_days);
+					start_ts = Timestamp::FromDatetime(start_date, start_time);
+				}
+			}
+
+			// Our interval difference will now give the correct result.
+			// Note that PG gives different interval subtraction results,
+			// so if we change this we will have to reimplement.
+			return Interval::GetDifference(end_ts, start_ts).months;
+		}
+	};
+
+	struct QuarterOperator {
+		template <class TA, class TB, class TR>
+		static inline TR Operation(TA start_ts, TB end_ts) {
+			return MonthOperator::Operation<TA, TB, TR>(start_ts, end_ts) / Interval::MONTHS_PER_QUARTER;
+		}
+	};
+	struct YearOperator {
+		template <class TA, class TB, class TR>
+		static inline TR Operation(TA start_ts, TB end_ts) {
+			return MonthOperator::Operation<TA, TB, TR>(start_ts, end_ts) / Interval::MONTHS_PER_YEAR;
+		}
+	};
+
+	struct DecadeOperator {
+		template <class TA, class TB, class TR>
+		static inline TR Operation(TA start_ts, TB end_ts) {
+			return MonthOperator::Operation<TA, TB, TR>(start_ts, end_ts) / Interval::MONTHS_PER_DECADE;
+		}
+	};
+
+	struct CenturyOperator {
+		template <class TA, class TB, class TR>
+		static inline TR Operation(TA start_ts, TB end_ts) {
+			return MonthOperator::Operation<TA, TB, TR>(start_ts, end_ts) / Interval::MONTHS_PER_CENTURY;
+		}
+	};
+
+	struct MilleniumOperator {
+		template <class TA, class TB, class TR>
+		static inline TR Operation(TA start_ts, TB end_ts) {
+			return MonthOperator::Operation<TA, TB, TR>(start_ts, end_ts) / Interval::MONTHS_PER_MILLENIUM;
+		}
+	};
+
+	struct DayOperator {
+		template <class TA, class TB, class TR>
+		static inline TR Operation(TA startdate, TB enddate) {
+			return (Timestamp::GetEpochNanoSeconds(enddate) - Timestamp::GetEpochNanoSeconds(startdate)) /
```

Comment:
Can we do these calculations with microseconds instead of with nanoseconds? The standard timestamp is stored in microseconds.

Reply:
Ah I was wondering about that.